### ISSUE TITLE
Build errors on Clang 3.7 because of new move diagnostics

### ISSUE THREAD

**Issue by dean0x7d** [dropbox/pyston#1016]
Trying to compile with Clang 3.7 results in the following errors because of `-Werror` and the new [move warnings](http://llvm.org/releases/3.7.0/tools/clang/docs/ReleaseNotes.html#improvements-to-clang-s-diagnostics) which are included with `-Wall` in the 3.7 release.

```
~/pyston_deps/llvm-trunk/include/llvm/ADT/APFloat.h:352:12: 
error: redundant move in return statement [-Werror,-Wredundant-move]
    return std::move(Value);
           ^
```

```
../../src/codegen/compvars.cpp:2548:16:
error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]
        return std::move(elts);
               ^
```
Note that the first error already has a [fix in LLVM](https://github.com/llvm-mirror/llvm/commit/fda9fd463dae38840cc02c5ae668cb1097395b9a).

**Comment by sYnfo** [comment]
:+1: I'm bumping into the same issue.

**Comment by undingen** [comment]
Could you please test if #1019 fixes the issue if you have time or if there are additional warnings which need to get fixed.
Thanks

**Comment by sYnfo** [comment]
Could be I'm doing something wrong, but the patch didn't seem to help:

https://gist.github.com/sYnfo/826bad71a20360700a3f

**Comment by sYnfo** [comment]
Unless I'm doing something wrong the patch doesn't seem to help:

https://gist.github.com/sYnfo/826bad71a20360700a3f

**Comment by undingen** [comment]
Ok I updated the patch to include fixes for the warnings in your log. Could you please try again?
If more problems show up I will install 3.7 :-D

**Comment by dean0x7d** [comment]
The patch does help: it clears the LLVM specific warnings.
But there are a few other `-Wredundant-move` warnings not related to LLVM and a large number of new `-Wshift-negative-value` messages show up now.

I've been looking into this and I can submit a pull request with the clang 3.7 fixes a bit later today.

**Comment by undingen** [comment]
I don't know if you have seen it but I replied a few seconds before your last replay :-D to let you know that I updated the patch to fix the remaining warnings (hope that those where the last ones..).

**Comment by dean0x7d** [comment]
Ah, I missed that. But there is actually a lot more. I have the fixes on my [fork](https://github.com/dean0x7d/pyston/commits/clang37_fixes).
There are 2 fixes for `-Wredundant-move`: 3205005f07941e442cda5d508802cac7b16c16ea
And 10 for `-Wshift-negative-value`: 60644abd67bff2e8c04675c3006f8db401af3043

That takes care of all the warnings, but then something truly weird happens. [The compiler crashes]( https://gist.github.com/dean0x7d/852456977ff50c726312) :-/
As far as I can tell the crash happens on these two lines with `delete obj;` [here](https://github.com/dropbox/pyston/blob/master/src/gc/gc.h#L230) and [here](https://github.com/dropbox/pyston/blob/master/src/gc/gc.h#L241). If I comment out those two lines, the build completes successfully.

I'm not sure if it's just something with my system or a genuine bug in Clang 3.7.

**Comment by kmod** [comment]
Ok I set up a container with clang-3.7 and I'll take a look :)

**Comment by kmod** [comment]
I'm running into the same clang bug -- I think this is a legit compiler bug so I [filed it on the clang tracker](https://llvm.org/bugs/show_bug.cgi?id=25700).  I think I found a workaround, and the build is succeeding in my container, so I opened #1020.  The tests aren't all passing though, but I'm guessing it's because I did this in a 15.10 container; I haven't looked into the test failures much (io_test is failing, set comparisons aren't being allowed in some places CPython is letting them, and stack_limits is failing).
